codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -334,9 +344,36 @@ class KotlinJmespathExpressionVisitor(
             arg.dotFunction(expression, "type()", ensureNullGuard = false)
         }
 
+        "sort" -> {
+            writer.addImport(RuntimeTypes.Core.Utils.jmespathSort)
+            val arg = expression.singleArg()
+            arg.dotFunction(expression, "jmespathSort()")
+        }
+
+        "sort_by" -> mappingFunction(expression, "sortBy", this::sortBy)
+
+        "max_by" -> mappingFunction(expression, "maxBy", this::maxBy)
+
+        "min_by" -> mappingFunction(expression, "minBy", this::minBy)
+
+        "map" -> mappingFunction(expression, "map", this::map, true)
+
         else -> throw CodegenException("Unknown function type in $expression")
     }
 
+    private fun mappingFunction(
+        expression: FunctionExpression,
+        variableName: String,
+        function: (JmespathExpression, VisitedExpression) -> String,
+        invertedArgs: Boolean = false,
+    ): VisitedExpression {
+        val (argIndex, expressionIndex) = if (invertedArgs) 1 to 0 else 0 to 1
```

Comment:
Can you explain why `map` has inverted arguments? I'm wondering if there is a "nicer" way to handle this?